ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/PerasCertDB/Impl.hs

```diff
@@ -0,0 +1,297 @@
+              , -- Note that the same block might be boosted by multiple points.
+                pvcsWeightByPoint =
+                  addToPerasWeightSnapshot boostedPt (getPerasCertBoost cert) pvcsWeightByPoint
+              , pvcsCertsByTicket =
+                  Map.insert pvcsLastTicketNo' cert pvcsCertsByTicket
+              , pvcsLastTicketNo = pvcsLastTicketNo'
+              }
+            (succ fp)
+        pure AddedPerasCertToDB
+  traceWith pcdbTracer $ case res of
+    AddedPerasCertToDB -> AddedPerasCert roundNo boostedPt
+    PerasCertAlreadyInDB -> IgnoredCertAlreadyInDB roundNo boostedPt
+  pure res
+ where
+  PerasCertDbEnv
+    { pcdbTracer
+    , pcdbVolatileState
+    } = env
+
+  boostedPt = getPerasCertBoostedBlock cert
+  roundNo = getPerasCertRound cert
+
+implGetWeightSnapshot ::
+  IOLike m =>
+  PerasCertDbEnv m blk -> STM m (WithFingerprint (PerasWeightSnapshot blk))
+implGetWeightSnapshot PerasCertDbEnv{pcdbVolatileState} =
+  fmap pvcsWeightByPoint <$> readTVar pcdbVolatileState
+
+implGetCertSnapshot ::
+  IOLike m =>
+  PerasCertDbEnv m blk -> STM m (PerasCertSnapshot blk)
+implGetCertSnapshot PerasCertDbEnv{pcdbVolatileState} =
+  readTVar pcdbVolatileState
+    <&> forgetFingerprint
+    <&> \PerasVolatileCertState
+           { pvcsCerts
+           , pvcsCertsByTicket
+           } ->
+        PerasCertSnapshot
+          { containsCert = \r -> Map.member r pvcsCerts
+          , getCertsAfter = \ticketNo ->
+              let (_, certs) = Map.split ticketNo pvcsCertsByTicket
+               in [(cert, tno) | (tno, cert) <- Map.toAscList certs]
+          }
+
+implGarbageCollect ::
+  forall m blk.
+  (IOLike m, StandardHash blk) =>
+  PerasCertDbEnv m blk -> SlotNo -> m ()
+implGarbageCollect PerasCertDbEnv{pcdbVolatileState} slot =
+  -- No need to update the 'Fingerprint' as we only remove certificates that do
+  -- not matter for comparing interesting chains.
+  atomically $ modifyTVar pcdbVolatileState (fmap gc)
+ where
+  gc :: PerasVolatileCertState blk -> PerasVolatileCertState blk
+  gc
+    PerasVolatileCertState
+      { pvcsCerts
+      , pvcsWeightByPoint
+      , pvcsLastTicketNo
+      , pvcsCertsByTicket
+      } =
+      PerasVolatileCertState
+        { pvcsCerts = Map.filter keepCert pvcsCerts
+        , pvcsWeightByPoint = prunePerasWeightSnapshot slot pvcsWeightByPoint
+        , pvcsCertsByTicket = Map.filter keepCert pvcsCertsByTicket
+        , pvcsLastTicketNo = pvcsLastTicketNo
+        }
+     where
+      keepCert cert =
+        pointSlot (getPerasCertBoostedBlock cert) >= NotOrigin slot
+
+{-------------------------------------------------------------------------------
+  Implementation-internal types
+-------------------------------------------------------------------------------}
+
+-- | Volatile Peras certificate state, i.e. certificates that could influence
+-- chain selection by boosting a volatile block.
+data PerasVolatileCertState blk = PerasVolatileCertState
```

Comment:
Perhaps we should add some invariant checking here. For instance:

- Should the different maps have all the same size? 
- For every certificate in `pvcsCerts`, is there a corresponding weight in `pvcsWeightByPoint`?
- Should the range of `pvcsCerts` and `pvcsCertsByTicket` be the same? 
- etc.